src/main/java/jcifs/smb/SmbTransportImpl.java

```diff
@@ -0,0 +1,1841 @@
+        }
+        else {
+            request.setResponse(response);
+        }
+        if ( response == null ) {
+            throw new IOException("Invalid response");
+        }
+
+        CommonServerMessageBlockRequest curHead = request;
+
+        int maxSize = getContext().getConfig().getMaximumBufferSize();
+
+        while ( curHead != null ) {
+            CommonServerMessageBlockRequest nextHead = null;
+            int totalSize = 0;
+            int n = 0;
+            CommonServerMessageBlockRequest last = null;
+            CommonServerMessageBlockRequest chain = curHead;
+            while ( chain != null ) {
+                n++;
+                int size = chain.size();
+                int cost = chain.getCreditCost();
+                CommonServerMessageBlockRequest next = chain.getNext();
+                if ( log.isTraceEnabled() ) {
+                    log.trace(
+                        String.format("%s costs %d avail %d (%s)", chain.getClass().getName(), cost, this.credits.availablePermits(), this.name));
+                }
+                if ( ( next == null || chain.allowChain(next) ) && totalSize + size < maxSize && this.credits.tryAcquire(cost) ) {
+                    totalSize += size;
+                    last = chain;
+                    chain = next;
+                }
+                else if ( last == null && totalSize + size > maxSize ) {
+                    throw new SmbException(String.format("Request size %d exceeds allowable size %d: %s", size, maxSize, chain));
+                }
+                else if ( last == null ) {
+                    // don't have enough credits/space for the first request, block until available
+                    // for space there is nothing we can do, callers need to make sure that a single message fits
+
+                    try {
+                        long timeout = getResponseTimeout(chain);
+                        if ( params.contains(RequestParam.NO_TIMEOUT) ) {
+                            this.credits.acquire(cost);
+                        }
+                        else {
+                            if ( !this.credits.tryAcquire(cost, timeout, TimeUnit.MILLISECONDS) ) {
+                                throw new SmbException("Failed to acquire credits in time");
+                            }
+                        }
+                        totalSize += size;
+                        // split off first request
+
+                        synchronized ( chain ) {
+                            CommonServerMessageBlockRequest snext = chain.split();
+                            nextHead = snext;
+                            if ( log.isDebugEnabled() && snext != null ) {
+                                log.debug("Insufficient credits, send only first " + chain + " next is " + snext);
+                            }
+                        }
+                        break;
+                    }
+                    catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)